File: scripts/targets/helix.ts
import { Palette, resolvePalettePath } from '../palette';

/**
 * Static syntax-role assignments. These reference palette *names* (resolved by
 * the [palette] block below), not colours, so they never change with the
 * source palette — only the named colours underneath them do.
 */
const ROLES = `keyword = { fg = "aqua" }
"keyword.control" = { fg = "aqua" }
"keyword.control.import" = { fg = "aqua" }
"keyword.control.return" = { fg = "aqua" }
"keyword.function" = { fg = "aqua" }
"keyword.operator" = { fg = "aqua" }
"keyword.directive" = { fg = "aqua" }
operator = { fg = "aqua" }
punctuation = { fg = "fg" }
"punctuation.delimiter" = { fg = "aqua" }
tag = { fg = "aqua" }
"constant.character.escape" = { fg = "aqua" }

function = { fg = "light-green" }
"function.builtin" = { fg = "cyan" }
"function.macro" = { fg = "light-green" }
attribute = { fg = "light-green", modifiers = ["italic"] }

string = { fg = "yellow" }
"string.regexp" = { fg = "yellow" }

type = { fg = "orange", modifiers = ["italic"] }
"type.builtin" = { fg = "orange", modifiers = ["italic"] }
constructor = { fg = "orange" }
"variable.parameter" = { fg = "orange", modifiers = ["italic"] }

constant = { fg = "purple" }
"constant.builtin" = { fg = "purple" }
"variable.builtin" = { fg = "cyan", modifiers = ["italic"] }
"variable.other.member" = { fg = "fg" }
namespace = { fg = "fg" }

"comment.block.documentation" = { fg = "comment", modifiers = ["italic"] }
"comment.line.documentation" = { fg = "comment", modifiers = ["italic"] }

"markup.heading" = { fg = "purple", modifiers = ["bold"] }
"markup.link" = { fg = "cyan", underline = { style = "line" } }
"markup.list" = { fg = "cyan" }
"markup.raw" = { fg = "light-green" }
"markup.quote" = { fg = "yellow", modifiers = ["italic"] }

"diff.plus" = { fg = "light-green" }
"diff.minus" = { fg = "red" }
"diff.delta" = { fg = "orange" }`;

/**
 * A [palette] entry: a named colour resolved from a palette path or a literal
 * `#rrggbb` (for Helix-specific shades the core palette doesn't carry), or a
 * blank line to preserve the grouping. `comment` is an optional inline note.
 */
type Entry = 'blank' | { name: string; ref: string; comment?: string };

/**
 * Helix's named colours. Shared identity colours point at the palette so they
 * track the source of truth; the rest are Helix-only shades kept as literals.
 */
const PALETTE: Entry[] = [
    { name: 'red', ref: 'syntax.error', comment: 'rust / logo-edge red' },
    { name: 'orange', ref: 'syntax.type', comment: 'logo orange — types, parameters' },
    { name: 'yellow', ref: 'syntax.string', comment: 'logo gold — strings' },
    { name: 'light-green', ref: 'syntax.function', comment: 'glowing hand green — functions' },
    { name: 'green', ref: '#9cbee6', comment: 'pale tube glow (unused member slot)' },
    { name: 'aqua', ref: 'syntax.keyword', comment: 'electric tube blue — keywords, operators' },
    { name: 'teal', ref: '#5f8fc4', comment: 'slate blue — markup, hints' },
    { name: 'turquoise', ref: '#9db4d8', comment: 'muted hangar blue' },
    { name: 'light-cyan', ref: '#c8e6ff', comment: 'ice glow' },
    { name: 'cyan', ref: 'syntax.builtin', comment: 'sky cyan — builtins' },
    { name: 'blue', ref: 'syntax.info', comment: 'bright tube blue — UI accents, labels' },
    { name: 'purple', ref: 'syntax.constant', comment: 'softened violet — constants, headings' },
    { name: 'magenta', ref: 'ansi.magenta', comment: 'sarcophagus pink — select mode' },
    { name: 'comment', ref: 'fg.comment', comment: 'muted blue shading' },
    { name: 'black', ref: 'bg.surface', comment: 'raised surface blue' },
    'blank',
    { name: 'add', ref: '#5cb454' },
    { name: 'change', ref: '#5878c4' },
    { name: 'delete', ref: '#a04a3a' },
    'blank',
    { name: 'error', ref: '#e05a3e', comment: 'deep rust' },
    { name: 'info', ref: '#55aef5' },
    { name: 'hint', ref: '#5f8fc4' },
    'blank',
    { name: 'fg', ref: 'fg.base', comment: 'pale blue highlight white' },
    { name: 'fg-dark', ref: 'fg.muted' },
    { name: 'fg-gutter', ref: '#33406f', comment: 'line numbers, whitespace marks' },
    { name: 'fg-linenr', ref: '#7484b8' },
    { name: 'fg-selected', ref: '#232d5c', comment: 'menu selection bg' },
    { name: 'border', ref: 'bg.border' },
    { name: 'border-highlight', ref: 'syntax.keyword' },
    { name: 'bg', ref: 'bg.base', comment: 'hangar deep blue' },
    { name: 'bg-inlay', ref: '#101d40' },
    { name: 'bg-selection', ref: 'bg.selection' },
    { name: 'bg-menu', ref: 'bg.sunken', comment: 'statusline, popups' },
    { name: 'bg-focus', ref: 'bg.overlay' },
];

const HEADER = `# rust-in-peace — generated from src/palette.json. Do not edit by hand.
inherits = "tokyonight"

# Syntax roles mirror the rust-in-peace VSCode theme (original structure):
#   keywords / operators / delimiters / tags = electric tube blue
#   functions / decorators = glow green      strings  = logo gold
#   types / parameters     = logo orange     constants = violet
#   builtins               = sky cyan        comments  = muted blue
# UI structure comes from the inherited base theme.`;

/** Generate the Helix TOML theme from the palette. */
export default (palette: Palette): string => {
    const width = Math.max(
        ...PALETTE.map(entry => (entry === 'blank' ? 0 : entry.name.length))
    );

    const paletteLines = PALETTE.map(entry => {
        if (entry === 'blank') {
            return '';
        }
        const hex = entry.ref.startsWith('#')
            ? entry.ref
            : resolvePalettePath(palette, entry.ref);
        const line = `${entry.name.padEnd(width)} = "${hex}"`;
        return entry.comment ? `${line} # ${entry.comment}` : line;
    });

    return [HEADER, '', ROLES, '', '[palette]', ...paletteLines, ''].join('\n');
};

File: scripts/targets/helix.py
"""Helix TOML theme.

A standalone theme (no `inherits`): syntax, markup, diagnostics, and the
full UI layer are all defined here, recoloured to the album palette and
mirroring the VS Code theme's token philosophy.

Design: the output has two layers. The ROLES block is static text — theme
roles reference palette *names* ("aqua", "orange"), so it never changes with
the source palette. The [palette] block beneath it defines those names, and
every one derives from the shared palette: identity colours resolve from
palette paths (diff and diagnostic slots take the VS Code theme's colour for
the equivalent UI element), while Helix-only shades are declared mix/adjust
formulas over palette anchors (scripts/color.py).
"""

from __future__ import annotations

from dataclasses import dataclass

from scripts.color import ColorFn, mixed
from scripts.palette import resolve_palette_path
from scripts.variants import Flavor

# Static: roles reference the palette names defined in the [palette] block below.
ROLES = """\
keyword = { fg = "aqua" }
"keyword.control" = { fg = "aqua" }
"keyword.control.import" = { fg = "aqua" }
"keyword.control.return" = { fg = "aqua" }
"keyword.function" = { fg = "aqua" }
"keyword.operator" = { fg = "aqua" }
"keyword.directive" = { fg = "aqua" }
operator = { fg = "aqua" }
punctuation = { fg = "fg" }
"punctuation.delimiter" = { fg = "aqua" }
tag = { fg = "aqua" }
special = { fg = "aqua" }
"constant.character.escape" = { fg = "aqua" }

function = { fg = "light-green" }
"function.builtin" = { fg = "cyan" }
"function.macro" = { fg = "light-green" }
"function.special" = { fg = "cyan" }
attribute = { fg = "light-green", modifiers = ["italic"] }

string = { fg = "yellow" }
"string.regexp" = { fg = "yellow" }
"string.special" = { fg = "aqua" }
"constant.character" = { fg = "yellow" }

type = { fg = "orange", modifiers = ["italic"] }
"type.builtin" = { fg = "orange", modifiers = ["italic"] }
"type.enum.variant" = { fg = "orange" }
constructor = { fg = "orange" }
"variable.parameter" = { fg = "orange", modifiers = ["italic"] }

constant = { fg = "purple" }
"constant.builtin" = { fg = "purple" }
variable = { fg = "fg" }
"variable.builtin" = { fg = "cyan", modifiers = ["italic"] }
"variable.other.member" = { fg = "fg" }
label = { fg = "blue" }
namespace = { fg = "fg" }

comment = { fg = "comment", modifiers = ["italic"] }
"comment.block.documentation" = { fg = "comment", modifiers = ["italic"] }
"comment.line.documentation" = { fg = "comment", modifiers = ["italic"] }

"markup.bold" = { modifiers = ["bold"] }
"markup.italic" = { modifiers = ["italic"] }
"markup.strikethrough" = { modifiers = ["crossed_out"] }
"markup.heading" = { fg = "purple", modifiers = ["bold"] }
"markup.heading.completion" = { bg = "bg-menu", fg = "fg" }
"markup.heading.hover" = { bg = "fg-selected" }
"markup.link" = { fg = "cyan", underline = { style = "line" } }
"markup.link.label" = { fg = "teal" }
"markup.link.text" = { fg = "teal" }
"markup.link.url" = { underline = { style = "line" } }
"markup.list" = { fg = "cyan" }
"markup.normal.completion" = { fg = "comment" }
"markup.normal.hover" = { fg = "fg-dark" }
"markup.raw" = { fg = "light-green" }
"markup.raw.inline" = { bg = "black", fg = "blue" }
"markup.quote" = { fg = "yellow", modifiers = ["italic"] }

"diff.plus" = { fg = "add" }
"diff.minus" = { fg = "delete" }
"diff.delta" = { fg = "change" }
"diff.delta.moved" = { fg = "blue" }

error = { fg = "error" }
warning = { fg = "yellow" }
info = { fg = "info" }
hint = { fg = "hint" }
"diagnostic.error" = { underline = { style = "curl", color = "error" } }
"diagnostic.warning" = { underline = { style = "curl", color = "yellow" } }
"diagnostic.info" = { underline = { style = "curl", color = "info" } }
"diagnostic.hint" = { underline = { style = "curl", color = "hint" } }
"diagnostic.unnecessary" = { modifiers = ["dim"] }
"diagnostic.deprecated" = { modifiers = ["crossed_out"] }

"ui.background" = { bg = "bg", fg = "fg" }
"ui.cursor" = { fg = "bg", bg = "fg-linenr" }
"ui.cursor.primary" = { fg = "bg", bg = "fg" }
"ui.cursor.primary.normal" = { fg = "bg", bg = "blue" }
"ui.cursor.primary.insert" = { fg = "bg", bg = "light-green" }
"ui.cursor.primary.select" = { fg = "bg", bg = "magenta" }
"ui.cursor.match" = { fg = "orange", modifiers = ["bold"] }
"ui.cursorline.primary" = { bg = "bg-inlay" }
"ui.help" = { bg = "bg-menu", fg = "fg" }
"ui.linenr" = { fg = "fg-gutter" }
"ui.linenr.selected" = { fg = "fg-linenr" }
"ui.menu" = { bg = "bg-menu", fg = "fg" }
"ui.menu.selected" = { bg = "fg-selected" }
"ui.popup" = { bg = "bg-menu", fg = "border-highlight" }
"ui.selection" = { bg = "bg-selection" }
"ui.selection.primary" = { bg = "bg-selection" }
"ui.statusline" = { bg = "bg-menu", fg = "fg-dark" }
"ui.statusline.inactive" = { bg = "bg-menu", fg = "fg-gutter" }
"ui.statusline.normal" = { bg = "blue", fg = "bg", modifiers = ["bold"] }
"ui.statusline.insert" = { bg = "light-green", fg = "bg", modifiers = ["bold"] }
"ui.statusline.select" = { bg = "magenta", fg = "bg", modifiers = ["bold"] }
"ui.text" = { fg = "fg" }
"ui.text.focus" = { bg = "bg-focus" }
"ui.text.inactive" = { fg = "comment", modifiers = ["italic"] }
"ui.text.info" = { bg = "bg-menu", fg = "fg" }
"ui.text.directory" = { fg = "cyan" }
"ui.virtual.ruler" = { bg = "fg-gutter" }
"ui.virtual.whitespace" = { fg = "fg-gutter" }
"ui.virtual.inlay-hint" = { bg = "bg-inlay", fg = "teal" }
"ui.virtual.jump-label" = { fg = "orange", modifiers = ["bold"] }
"ui.window" = { fg = "border", modifiers = ["bold"] }"""


@dataclass(frozen=True)
class Entry:
    """A [palette] entry.

    A named colour resolved from a palette path, or derived from palette
    anchors via a ColorFn (scripts/color.py) for Helix-specific shades the
    core palette doesn't carry directly.
    """

    name: str
    ref: str | ColorFn
    comment: str | None = None


BLANK = None  # a blank line, preserving the grouping in the output

# Diagnostic hints and the teal slot share one slate blue: info sunk toward
# the selection surface.
SLATE = mixed("bg.selection", "syntax.info", 0.70)

# Helix's named colours, grouped as in the output (BLANK = blank line).
# Slots with a VS Code UI equivalent take that element's colour; the rest
# derive from the semantically-nearest palette anchors.
PALETTE: list[Entry | None] = [
    Entry("orange", "syntax.type", "logo orange — types, parameters"),
    Entry("yellow", "syntax.string", "logo gold — strings"),
    Entry("light-green", "syntax.function", "glowing hand green — functions"),
    Entry("aqua", "syntax.keyword", "electric tube blue — keywords, operators"),
    Entry("teal", SLATE, "slate blue — markup, hints"),
    Entry("cyan", "syntax.builtin", "sky cyan — builtins"),
    Entry("blue", "syntax.info", "bright tube blue — UI accents, labels"),
    Entry("purple", "syntax.constant", "softened violet — constants, headings"),
    Entry("magenta", "ansi.magenta", "sarcophagus pink — select mode"),
    Entry("comment", "fg.comment", "muted blue shading"),
    Entry("black", "bg.surface", "raised surface blue"),
    BLANK,
    Entry("add", "syntax.function", "as VS Code gitDecoration.added"),
    Entry("change", "syntax.builtin", "as VS Code gitDecoration.modified"),
    Entry("delete", "syntax.error", "as VS Code gitDecoration.deleted"),
    BLANK,
    Entry("error", "syntax.error", "as VS Code editorError"),
    Entry("info", "syntax.builtin", "as VS Code editorInfo"),
    Entry("hint", SLATE),
    BLANK,
    Entry("fg", "fg.base", "pale blue highlight white"),
    Entry("fg-dark", "fg.muted"),
    Entry(
        "fg-gutter",
        mixed("bg.base", "fg.ink", 0.10),
        "whitespace marks — flattened editorWhitespace",
    ),
    Entry("fg-linenr", "fg.comment", "as VS Code editorLineNumber"),
    Entry("fg-selected", "bg.selection", "menu selection bg — as VS Code list selection"),
    Entry("border", "bg.border"),
    Entry("border-highlight", "syntax.keyword"),
    Entry("bg", "bg.base", "hangar deep blue"),
    Entry("bg-inlay", mixed("bg.base", "bg.selection", 0.50), "flattened list.focusBackground"),
    Entry("bg-selection", "bg.selection"),
    Entry("bg-menu", "bg.sunken", "statusline, popups"),
    Entry("bg-focus", "bg.overlay"),
]

HEADER = """\
# {slug} — a Helix theme from the Megadeth "Rust in Peace" cover palette.
# Generated from the src/ palettes by scripts/targets/helix.py (just build-themes).
# Do not edit by hand: edit the palette and rebuild.

# Syntax roles mirror the rust-in-peace VSCode theme (original structure):
#   keywords / operators / delimiters / tags = electric tube blue
#   functions / decorators = glow green      strings  = logo gold
#   types / parameters     = logo orange     constants = violet
#   builtins               = sky cyan        comments  = muted blue
# The theme is standalone — markup, diagnostics, and the full UI layer are
# defined below. Every colour in [palette] derives from the shared source of
# truth: identity slots map straight to it (diffs and diagnostics follow the
# VSCode theme's UI elements), Helix-only shades via declared mix formulas."""


def generate(flavor: Flavor) -> str:
    """Generate the Helix TOML theme for one flavor."""
    palette = flavor.palette
    width = max(len(entry.name) for entry in PALETTE if entry is not None)

    palette_lines = []
    for entry in PALETTE:
        if entry is None:
            palette_lines.append("")
            continue
        ref = entry.ref
        hex_colour = ref(palette) if callable(ref) else resolve_palette_path(palette, ref)
        line = f'{entry.name.ljust(width)} = "{hex_colour}"'
        palette_lines.append(f"{line} # {entry.comment}" if entry.comment else line)

    header = HEADER.format(slug=flavor.slug)
    return "\n".join([header, "", ROLES, "", "[palette]", *palette_lines, ""])
